ariadne/src/Ariadne/Wallet/UiAdapter.hs

```diff
@@ -0,0 +1,133 @@
+module Ariadne.Wallet.UiAdapter
+  ( UiWalletData(..)
+  , UiAccountData(..)
+  , UiWalletSelection(..)
+  -- * Lens
+  , uwdAccounts
+  , uadAddresses
+  --
+  , toUiWalletDatas
+  , toUiWalletSelection
+  ) where
+
+import Universum
+
+import qualified Data.Vector as V
+
+import Ariadne.Wallet.Cardano.Kernel.DB.AcidState
+import Ariadne.Wallet.Cardano.Kernel.DB.HdWallet
+import Ariadne.Wallet.Cardano.Kernel.DB.HdWallet.Read
+import Ariadne.Wallet.Cardano.Kernel.DB.InDb
+import Ariadne.Wallet.Cardano.Kernel.DB.Util.IxSet
+import Ariadne.Wallet.Face
+
+import Control.Lens (makeLenses)
+import Data.List.Index (indexed)
+import Safe (headMay)
+
+-- 'WalletData' like data type, used only in UI glue
+data UiWalletData = UiWalletData
+  { _uwdName     :: !Text
+  , _uwdId       :: !HdRootId
+  , _uwdAccounts :: !(Vector UiAccountData)
+  } deriving (Show, Generic)
+
+data UiAccountData = UiAccountData
+  { _uadName      :: !Text
+  , _uadPath      :: !Word32
+  -- Because of ChainType layer Now it should be ((HdAddressChain, Word32), Address) I guess, but
+  -- AFAIU we don't want a new layer, so addresses of both types wiil be in one list -- External first.
+  , _uadAddresses :: !(Vector (Word32, Address))
```

Comment:
BTW, can you clarify what does this `Word32` mean? Is it unique? I.e., can I use at key in map to store addresses of an account?

Reply:
@maksbotan `Word32` in `_uadPath` is just an index of account in a list (list representation of `IxSet HdAccount`) of wallet's accounts.
`Word32` in `_uadAddresses` is just an index of address in a list of account's addresses. Note, that these indexes are not stable -- removing an account or an address cause changes in indexation. `HdAccountIdIx` and `HdAddressIdIx` are stable ones.